base/broadcast.jl

```diff
@@ -257,13 +257,23 @@ end
 @inline broadcast_elwise_op(f, As...) =
     broadcast!(f, similar(Array{promote_eltype_op(f, As...)}, broadcast_indices(As...)), As...)
 
-@pure typestuple(a) = Tuple{eltype(a)}
-@pure typestuple(T::Type) = Tuple{Type{T}}
-@pure typestuple(a, b...) = Tuple{typestuple(a).types..., typestuple(b...).types...}
+typestuple(a) = (Base.@_pure_meta; Tuple{eltype(a)})
+typestuple(T::Type) = (Base.@_pure_meta; Tuple{Type{T}})
+typestuple(a, b...) = (Base.@_pure_meta; Tuple{typestuple(a).types..., typestuple(b...).types...})
+
+ftype(f, A) = typeof(a->f(a))
```

Comment:
We need to handle the cases `ftype(f, A...)` and `ftype(T::Type, A...)` separately anyway. Otherwise, `Core.Typeof` would be have been a would option for `ftype(f, A) = typeof(f)` and `ftype(T::Type, A) = Type{T}`